---
title: "Introgress"
author: "Emily Ostrow"
date: "10/19/2021"
output: pdf_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(introgress)
library(vcfR)
library(adegenet)
library(pophelper)

vcf <- read.vcfR("m3MAC2Unlinkgq30.vcf.gz")
gen<-vcfR2genlight(vcf)
colnames(vcf@gt)
#locs <- read.csv("vcfLatLong.csv", stringsAsFactors = F)
dim(vcf)
sfiles <- "structureOutk3v1.txt"
slist <- readQ(files=sfiles)

##read in points and vcf
points <- read.csv("StructureOrderPoints.csv", stringsAsFactors = F, header = T)
allindsvcf <- read.vcfR("m3MAC2Unlinkgq30.vcf.gz")

###identify which individuals are 90% or greater in a population
structurePops <- cbind(slist$structureOutk3v1$Cluster1,slist$structureOutk3v1$Cluster2,slist$structureOutk3v1$Cluster3,rep(4, times=length(colnames(vcf@gt))-1), rep(5, times=length(colnames(vcf@gt))-1))
row.names(structurePops) <- points$LongName
clust1Vec90 <- structurePops[,1]>=.9
clust2Vec90 <- structurePops[,2]>=.9
clust3Vec90 <- structurePops[,3]>=.9

clust1Names <- row.names(structurePops[clust1Vec90,])
clust2Names <- row.names(structurePops[clust2Vec90,])
clust3Names <- row.names(structurePops[clust3Vec90,])

###identifying which structure population each individual most closely aligns with
for (i in 1:length(structurePops[,1])){
  if (structurePops[i,1]>structurePops[i,2]&&structurePops[i,1]>structurePops[i,3]){
    structurePops[i,4] <- 1
  }
  else if (structurePops[i,2]>structurePops[i,1]&&structurePops[i,2]>structurePops[i,3]){
    structurePops[i,4] <- 2
  }
  else{
    structurePops[i,4] <- 3
  }
}
###same as above but creating a category for inds that have large portions of all pops
for (i in 1:length(structurePops[,1])){
  if (structurePops[i,1]<0.5&&structurePops[i,2]<0.5&&structurePops[i,3]<0.5){
    structurePops[i,5] <- 4
  }
  else if (structurePops[i,1]>structurePops[i,2]&&structurePops[i,1]>structurePops[i,3]){
    structurePops[i,5] <- 1
  }
  else if (structurePops[i,2]>structurePops[i,1]&&structurePops[i,2]>structurePops[i,3]){
    structurePops[i,5] <- 2
  }
  else{
    structurePops[i,5] <- 3
  }
}

###create lists of inds with majority in a pop and create a vcf for those comparisons
compareEastNorthinds <- c(row.names(structurePops[structurePops[,4]==1,]),row.names(structurePops[structurePops[,4]==2,]))
compareEastSouthinds <- c(row.names(structurePops[structurePops[,4]==1,]),row.names(structurePops[structurePops[,4]==3,]))
compareNorthSouthinds <- c(row.names(structurePops[structurePops[,4]==2,]),row.names(structurePops[structurePops[,4]==3,]))
EastNorthVcf <- vcf
EastSouthVcf <- vcf
NorthSouthVcf <- vcf
EastNorthVcf@gt <- EastNorthVcf@gt[,colnames(EastNorthVcf@gt) %in% c("FORMAT",compareEastNorthinds)]
EastSouthVcf@gt <- EastSouthVcf@gt[,colnames(EastSouthVcf@gt) %in% c("FORMAT",compareEastSouthinds)]
NorthSouthVcf@gt <- NorthSouthVcf@gt[,colnames(NorthSouthVcf@gt) %in% c("FORMAT",compareNorthSouthinds)]
EastNorthGen <- vcfR2genlight(EastNorthVcf)
EastSouthGen <- vcfR2genlight(EastSouthVcf)
NorthSouthGen<- vcfR2genlight(NorthSouthVcf)
```

# EastSouth Comparison

```{r EastSouth, warning=F, message=F, cache=T }
currentVCF <- EastSouthVcf
currentGen <- EastSouthGen

#create SNP matrices
mat<-extract.gt(currentVCF)
dim(mat)
for (i in 1:nrow(mat)){
  for (j in 1:ncol(mat)){
    mat[i,j] <- paste0(substr(mat[i,j], 1,1), substr(mat[i,j], 3, 3))
    j <- j+1
  }
  i <- i+1
}

mat[1:5,1:5]

conv.mat<-mat
conv.mat[conv.mat == "00"]<-0
conv.mat[conv.mat == "01"]<-1
conv.mat[conv.mat == "11"]<-2
conv.mat[conv.mat == "NANA"]<-NA
conv.mat<-as.data.frame(conv.mat)
#convert to numeric
for (i in 1:ncol(conv.mat)){
  conv.mat[,i]<-as.numeric(as.character(conv.mat[,i]))
}
conv.mat2 <- conv.mat
conv.mat2[conv.mat2 == 0]<-as.character("0/0")
conv.mat2[conv.mat2 == 1]<-as.character("0/1")
conv.mat2[conv.mat2 == 2]<-as.character("1/1")


#calc AF for the samples you will use to call fixed differences
east<-(rowSums(conv.mat[,c("B_virginianus_ENO319", "B_virginianus_ENO309", "B_virginianus_159674")], na.rm=T)/(rowSums(is.na(conv.mat[,c("B_virginianus_ENO319", "B_virginianus_ENO309", "B_virginianus_159674")]) == FALSE)))/2
south<-(rowSums(conv.mat[,c("B_virginianus_18275", "B_virginianus_23829", "B_virginianus_100443")], na.rm=T)/(rowSums(is.na(conv.mat[,c("B_virginianus_18275", "B_virginianus_23829", "B_virginianus_100443")]) == FALSE)))/2

#find fixed SNPs
diff<-abs(east - south)
#how many SNPs are fixed
table(is.na(diff) == FALSE & diff > .8)



#subsample matrix converted for AF calcs to only fixed SNPS
SNPsitesFilter <- row.names(conv.mat[is.na(diff) == FALSE & diff > .8,])
conv.mat<-conv.mat[is.na(diff) == FALSE & diff > .8,]
dim(conv.mat)
conv.mat2 <- conv.mat2[SNPsitesFilter,]
dim(conv.mat2)
#write a logical test to convert alleles so that a single number represents one parental ancestry
for (i in 1:nrow(conv.mat2)){
  ###list 3 inds from one pop
  if(sum(conv.mat[i,c("B_virginianus_ENO319", "B_virginianus_ENO309", "B_virginianus_159674")], na.rm = T)/3>=1){
    for (j in 1:ncol(conv.mat2)){
      if (is.na(conv.mat2[i,j])){
        j <- j+1 
      }
        else if(conv.mat2[i,j]=="1/1"){
        conv.mat2[i,j][conv.mat2[i,j] == "1/1"]<-"0/0"
        j <- j+1  
      }else if(conv.mat2[i,j]=="0/0"){
        conv.mat2[i,j][conv.mat2[i,j] == "0/0"]<-"1/1"
        j <- j+1  
      }else{
        j <- j+1  
      }
      
    }
  }
}
conv.mat2[is.na(conv.mat2) ]<-"NA/NA"


#make locus info df
locus.info<-data.frame(locus=rownames(conv.mat2),
                       type=rep("C", times=nrow(conv.mat2)),
                       lg=currentVCF@fix[,1][is.na(diff) == FALSE & diff > .8],
                       marker.pos=currentVCF@fix[,2][is.na(diff) == FALSE & diff > .8])

#make bpcum
nCHR <- length(unique(locus.info$lg))
locus.info$BPcum <- NA
s <- 0
nbp <- c()


#we now have a gt matrix in proper format for introgress
#convert genotype data into a matrix of allele counts
count.matrix<-prepare.data(admix.gen=conv.mat2, loci.data=locus.info,
                           parental1="1",parental2="0", pop.id=F,
                           ind.id=F, fixed=T)

#estimate hybrid index values
hi.index.sim<-est.h(introgress.data=count.matrix,loci.data=locus.info,
                    fixed=T, p1.allele="1", p2.allele="0")

locus.info$locus<-rep("", times=nrow(locus.info))
#LociDataSim1$lg<-c(1:110)
mk.image(introgress.data=count.matrix, loci.data=locus.info,
         marker.order=order(locus.info$BPcum),hi.index=hi.index.sim, ylab.image="Individuals",
         xlab.h="population 2 ancestry", pdf=F,
         col.image=c(rgb(1,0,0,alpha=.5),rgb(0,0,0,alpha=.8),rgb(0,0,1,alpha=.5)))
#calculate mean heterozygosity across these 110 fixed markers for each sample
#using their function
het<-calc.intersp.het(introgress.data=count.matrix)

#plot triangle
##eastSouth90pcolor
plot(x=hi.index.sim$h, y=het, bg=c(rep(rgb(1,1,1,alpha=.5), times=8),rep(rgb(1,0,0,alpha=.5), times=10),
                                   rep(rgb(1,1,1,alpha=.5), times=2), rep(rgb(1,0,0,alpha=.5), times=2),
                                   rep(rgb(1,1,1,alpha=.5), times=3), rep(rgb(1,0,0,alpha=.5), times=1),
                                   rep(rgb(1,1,1,alpha=.5), times=1), rep(rgb(1,0,0,alpha=.5), times=2),
                                   rep(rgb(1,1,1,alpha=.5), times=1), rep(rgb(1,0,0,alpha=.5), times=1),
                                   rep(rgb(1,1,1,alpha=.5), times=6), rep(rgb(0,0,1,alpha=.5), times=1),
                                   rep(rgb(1,1,1,alpha=.5), times=2), rep(rgb(0,0,1,alpha=.5), times=1),
                                   rep(rgb(1,1,1,alpha=.5), times=10), rep(rgb(1,0,0,alpha=.5), times=1),
                                   rep(rgb(1,1,1,alpha=.5), times=6), rep(rgb(0,0,1,alpha=.5), times=1),
                                   rep(rgb(1,1,1,alpha=.5), times=1), rep(rgb(1,0,0,alpha=.5), times=1)),
     pch=21, cex=1.5,
     xlab="Hybrid Index", ylab="Heterozygosity",
     ylim=c(0,1), xlim=c(0,1))
segments(x0 =0, y0 =0, x1 =.5, y1 =1)
segments(x0 =1, y0 =0, x1 =.5, y1 =1)

#dev.off()

```


\newpage
# EastNorth Comparison

```{r EastNorth, warning=F, message=F, cache=T }
currentVCF <- EastNorthVcf
currentGen <- EastNorthGen

#create SNP matrices
mat<-extract.gt(currentVCF)
dim(mat)
for (i in 1:nrow(mat)){
  for (j in 1:ncol(mat)){
    mat[i,j] <- paste0(substr(mat[i,j], 1,1), substr(mat[i,j], 3, 3))
    j <- j+1
  }
  i <- i+1
}

mat[1:5,1:5]

conv.mat<-mat
conv.mat[conv.mat == "00"]<-0
conv.mat[conv.mat == "01"]<-1
conv.mat[conv.mat == "11"]<-2
conv.mat[conv.mat == "NANA"]<-NA
conv.mat<-as.data.frame(conv.mat)
#convert to numeric
for (i in 1:ncol(conv.mat)){
  conv.mat[,i]<-as.numeric(as.character(conv.mat[,i]))
}
conv.mat2 <- conv.mat
conv.mat2[conv.mat2 == 0]<-as.character("0/0")
conv.mat2[conv.mat2 == 1]<-as.character("0/1")
conv.mat2[conv.mat2 == 2]<-as.character("1/1")


#calc AF for the samples you will use to call fixed differences
east<-(rowSums(conv.mat[,c("B_virginianus_ENO319", "B_virginianus_48729", "B_virginianus_36185")], na.rm=T)/(rowSums(is.na(conv.mat[,c("B_virginianus_ENO319", "B_virginianus_48729", "B_virginianus_36185")]) == FALSE)))/2
north<-(rowSums(conv.mat[,c("B_virginianus_53496", "B_virginianus_44278", "B_virginianus_JJW5053")], na.rm=T)/(rowSums(is.na(conv.mat[,c("B_virginianus_53496", "B_virginianus_44278", "B_virginianus_JJW5053")]) == FALSE)))/2

#find fixed SNPs
diff<-abs(east - north)
#how many SNPs are fixed
table(is.na(diff) == FALSE & diff > .8)


currentVCF@fix[,1][is.na(diff) == FALSE & diff > .8]


#subsample matrix converted for AF calcs to only fixed SNPS
SNPsitesFilter <- row.names(conv.mat[is.na(diff) == FALSE & diff > .8,])
conv.mat<-conv.mat[is.na(diff) == FALSE & diff > .8,]
dim(conv.mat)
conv.mat2 <- conv.mat2[SNPsitesFilter,]
dim(conv.mat2)
#write a logical test to convert alleles so that a single number represents one parental ancestry
for (i in 1:nrow(conv.mat2)){
  ###list 3 inds from one pop
  if(sum(conv.mat[i,c("B_virginianus_53496", "B_virginianus_44278", "B_virginianus_JJW5053")], na.rm = T)/3>=1){
    for (j in 1:ncol(conv.mat2)){
      if (is.na(conv.mat2[i,j])){
        j <- j+1 
      }
        else if(conv.mat2[i,j]=="1/1"){
        conv.mat2[i,j][conv.mat2[i,j] == "1/1"]<-"0/0"
        j <- j+1  
      }else if(conv.mat2[i,j]=="0/0"){
        conv.mat2[i,j][conv.mat2[i,j] == "0/0"]<-"1/1"
        j <- j+1  
      }else{
        j <- j+1  
      }
      
    }
  }
}
conv.mat2[is.na(conv.mat2) ]<-"NA/NA"


#make locus info df
locus.info<-data.frame(locus=rownames(conv.mat2),
                       type=rep("C", times=nrow(conv.mat2)),
                       lg=currentVCF@fix[,1][is.na(diff) == FALSE & diff > .8],
                       marker.pos=currentVCF@fix[,2][is.na(diff) == FALSE & diff > .8])

#make bpcum
nCHR <- length(unique(locus.info$lg))
locus.info$BPcum <- NA
s <- 0
nbp <- c()


#we now have a gt matrix in proper format for introgress
#convert genotype data into a matrix of allele counts
count.matrix<-prepare.data(admix.gen=conv.mat2, loci.data=locus.info,
                           parental1="1",parental2="0", pop.id=F,
                           ind.id=F, fixed=T)

#estimate hybrid index values
hi.index.sim<-est.h(introgress.data=count.matrix,loci.data=locus.info,
                    fixed=T, p1.allele="1", p2.allele="0")

locus.info$locus<-rep("", times=nrow(locus.info))
#LociDataSim1$lg<-c(1:110)
mk.image(introgress.data=count.matrix, loci.data=locus.info,
         marker.order=order(locus.info$BPcum),hi.index=hi.index.sim, ylab.image="Individuals",
         xlab.h="population 2 ancestry", pdf=F,
         col.image=c(rgb(1,0,0,alpha=.5),rgb(0,0,0,alpha=.8),rgb(0,0,1,alpha=.5)))
#calculate mean heterozygosity across these 110 fixed markers for each sample
#using their function
het<-calc.intersp.het(introgress.data=count.matrix)

#plotTriangle
##eastnorth90pcolor
plot(x=hi.index.sim$h, y=het, bg=c(rep(rgb(1,0,0,alpha=.5), times=21),rep(rgb(1,1,1,alpha=.5), times=4),
                                        rep(rgb(1,0,0,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=1),
                                        rep(rgb(1,0,0,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=5),
                                        rep(rgb(0,0,1,alpha=.5), times=10),rep(rgb(1,1,1,alpha=.5), times=2),
                                        rep(rgb(0,0,1,alpha=.5), times=2),rep(rgb(1,1,1,alpha=.5), times=4),
                                        rep(rgb(0,0,1,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=1),
                                        rep(rgb(0,0,1,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=1),
                                        rep(rgb(0,0,1,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=1),
                                        rep(rgb(0,0,1,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=16),
                                        rep(rgb(0,0,1,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=11),
                                        rep(rgb(0,0,1,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=4)),
     pch=21, cex=1.5,
     xlab="Hybrid Index", ylab="Heterozygosity",
     ylim=c(0,1), xlim=c(0,1))
segments(x0 =0, y0 =0, x1 =.5, y1 =1)
segments(x0 =1, y0 =0, x1 =.5, y1 =1)

#dev.off()
```


\newpage
# NorthSouth Comparison

```{r NorthSouth, warning=F, message=F, cache=T }
currentVCF <- NorthSouthVcf
currentGen <- NorthSouthGen
structurePops[compareNorthSouthinds,1:5] 




#create SNP matrices
mat<-extract.gt(currentVCF)
dim(mat)
for (i in 1:nrow(mat)){
  for (j in 1:ncol(mat)){
    mat[i,j] <- paste0(substr(mat[i,j], 1,1), substr(mat[i,j], 3, 3))
    j <- j+1
  }
  i <- i+1
}

mat[1:5,1:5]

conv.mat<-mat
conv.mat[conv.mat == "00"]<-0
conv.mat[conv.mat == "01"]<-1
conv.mat[conv.mat == "11"]<-2
conv.mat[conv.mat == "NANA"]<-NA
conv.mat<-as.data.frame(conv.mat)
#convert to numeric
for (i in 1:ncol(conv.mat)){
  conv.mat[,i]<-as.numeric(as.character(conv.mat[,i]))
}
conv.mat2 <- conv.mat
conv.mat2[conv.mat2 == 0]<-"0/0"
conv.mat2[conv.mat2 == 1]<-"0/1"
conv.mat2[conv.mat2 == 2]<-"1/1"


#calc AF for the samples you will use to call fixed differences
north<-(rowSums(conv.mat[,c("B_virginianus_UAMX6949", "B_virginianus_UAMX6204", "B_virginianus_JJW5053")], na.rm=T)/(rowSums(is.na(conv.mat[,c("B_virginianus_UAMX6949", "B_virginianus_UAMX6204", "B_virginianus_JJW5053")]) == FALSE)))/2
south<-(rowSums(conv.mat[,c("B_virginianus_28702", "B_virginianus_18275", "B_virginianus_21020")], na.rm=T)/(rowSums(is.na(conv.mat[,c("B_virginianus_28702", "B_virginianus_18275", "B_virginianus_21020")]) == FALSE)))/2
###fixed SNPs from Fst analyses
fixedNorthSouthSNPs <- c("999_25", "1446_60", "2272_73", "2278_42", "3058_79", "4010_28", "4442_45", "5076_39", "5539_57", "6070_60", "6408_63", "8301_22", "9341_33", "9764_38", "12024_30", "12826_16", "13154_64", "14270_42", "21508_37", "21631_50", "22030_7", "22904_48", "23070_60", "23421_12", "24632_22", "25459_43", "27288_77", "27481_70", "29415_90", "29555_28", "30420_29", "30461_52", "33355_43", "33822_34", "34455_63", "34515_33", "35508_69", "35982_25", "36147_21", "36204_23", "37716_55", "38176_24", "38265_38", "39061_7", "41607_41")

#find fixed SNPs
#subsample original matrix to only fixed diff SNPs
fixedSnpVec <- vector()
for (i in 1:length(rownames(mat))){
  for (j in 1:length(fixedNorthSouthSNPs)){
    if(rownames(mat)[i]!=fixedNorthSouthSNPs[j]){
      j <- j+1
    }
    else{
      fixedSnpVec <- c(fixedSnpVec,i)
      cat("found one: row in matrix is:", i, "and position in vector is:", j, "\n")
      j <- length(fixedNorthSouthSNPs)
      
    }
    i <- i+1
  }
}



conv.mat<-conv.mat[fixedSnpVec,]
dim(conv.mat)
conv.mat2 <- conv.mat2[fixedSnpVec,]
dim(conv.mat2)

#write a logical test to convert alleles so that a single number represents one parental ancestry
for (i in 1:nrow(conv.mat2)){
  ###list 3 inds from one pop
  if(sum(conv.mat[i,c(2, 5, 8)], na.rm = T)/3>=1){
    for (j in 1:ncol(conv.mat2)){
      if (is.na(conv.mat2[i,j])){
        j <- j+1 
      }
        else if(conv.mat2[i,j]=="1/1"){
        conv.mat2[i,j][conv.mat2[i,j] == "1/1"]<-"0/0"
        j <- j+1  
      }else if(conv.mat2[i,j]=="0/0"){
        conv.mat2[i,j][conv.mat2[i,j] == "0/0"]<-"1/1"
        j <- j+1  
      }else{
        j <- j+1  
      }
      
    }
  }
}
conv.mat2[is.na(conv.mat2) ]<-"NA/NA"


#make locus info df
locus.info<-data.frame(locus=rownames(conv.mat2),
                       type=rep("C", times=nrow(conv.mat2)),
                       lg=currentVCF@fix[,1][fixedSnpVec],
                       marker.pos=currentVCF@fix[,2][fixedSnpVec])

#make bpcum
nCHR <- length(unique(locus.info$lg))
locus.info$BPcum <- NA
s <- 0
nbp <- c()


#we now have a gt matrix in proper format for introgress
#convert genotype data into a matrix of allele counts
count.matrix<-prepare.data(admix.gen=conv.mat2, loci.data=locus.info,
                           parental1="1",parental2="0", pop.id=F,
                           ind.id=F, fixed=T)

#estimate hybrid index values
hi.index.sim<-est.h(introgress.data=count.matrix,loci.data=locus.info,
                    fixed=T, p1.allele="1", p2.allele="0")

locus.info$locus<-rep("", times=nrow(locus.info))
mk.image(introgress.data=count.matrix, loci.data=locus.info,
         marker.order=order(locus.info$BPcum),hi.index=hi.index.sim, ylab.image="Individuals",
         xlab.h="population 2 ancestry", pdf=F,
         col.image=c(rgb(1,0,0,alpha=.5),rgb(0,0,0,alpha=.8),rgb(0,0,1,alpha=.5)))

#calculate mean heterozygosity across these 110 fixed markers for each sample
het<-calc.intersp.het(introgress.data=count.matrix)

#plot triangle
##northsouth90pcolor
     plot(x=hi.index.sim$h, y=het, bg=c(rep(rgb(1,0,0,alpha=.5), times=21),rep(rgb(1,1,1,alpha=.5), times=8),
                                        rep(rgb(1,0,0,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=2),
                                        rep(rgb(1,0,0,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=12),
                                        rep(rgb(0,0,1,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=4),
                                        rep(rgb(0,0,1,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=17),
                                        rep(rgb(0,0,1,alpha=.5), times=1),rep(rgb(1,1,1,alpha=.5), times=7)),
          

     
     pch=21, cex=1.5,
     xlab="Hybrid Index", ylab="Heterozygosity",
     ylim=c(0,1))
segments(x0 =0, y0 =0, x1 =.5, y1 =1)
segments(x0 =1, y0 =0, x1 =.5, y1 =1)

#dev.off()


```
